Win32-network/src/System/Win32/NamedPipes.hsc

```diff
@@ -191,3 +197,53 @@ foreign import ccall safe "windows.h WaitNamedPipeA"
   c_WaitNamedPipe :: LPCSTR -- lpNamedPipeName
                   -> DWORD  -- nTimeOut
                   -> IO Bool
+
+
+-- | A reliable connect call, as designed in
+-- <https://docs.microsoft.com/en-us/windows/win32/ipc/named-pipe-client>
+--
+-- The arguments are passed directly to 'createFile'.
+--
+-- Note we pick the more familiar posix naming convention, do not confuse this
+-- function with 'connectNamedPipe' (which corresponds to posix 'accept')
+--
+connect :: String                      -- ^ file name
+        -> AccessMode                  -- ^ dwDesiredAccess
+        -> ShareMode                   -- ^ dwSharedMode
+        -> Maybe LPSECURITY_ATTRIBUTES -- ^ lpSecurityAttributes
+        -> CreateMode                  -- ^ dwCreationDisposition
+        -> FileAttributeOrFlag         -- ^ dwFlagsAndAttributes
+        -> Maybe HANDLE                -- ^ hTemplateFile
+        -> IO HANDLE
+connect fileName dwDesiredAccess dwSharedMode lpSecurityAttributes dwCreationDisposition dwFlagsAndAttributes hTemplateFile = connectLoop
+  where
+    connectLoop = do
+      -- `createFile` checks for `INVALID_HANDLE_VALUE` and retries if this is
+      -- caused by `ERROR_SHARING_VIOLATION`.
+      mh <- try $
+              createFile fileName
+                         dwDesiredAccess
+                         dwSharedMode
+                         lpSecurityAttributes
+                         dwCreationDisposition
+                         dwFlagsAndAttributes
+                         hTemplateFile
+      case mh :: Either IOException HANDLE of
+        Left e -> do
+          -- unfortuntally 'ERROR_PIPE_BUSY' is not part of 'errentry' in
+          -- `libraries/cbits/Win32Utils.c`.
+          errorCode <- getLastError
+          when (errorCode /= eRROR_PIPE_BUSY)
+            $ throwIO e
+          -- all pipe instance where busy, wait 20s and retry;
+          waitNamedPipe fileName 20_000
```

Comment:
But `waitNamedPipe` throws exceptions on timeout.

We should either have `waitNamedPipe` return a result and not throw, or we should catch the timeout result here and ignore it. It's ok if it times out, we just go round again.

Is 20s too long? We don't know if this would block shutdown until it times out. If not it's ok.

Reply:
> But waitNamedPipe throws exceptions on timeout.

Well if we cannot connect in `20s` we surely should throw an exception. But if we want to make this loop tighter, and wait at most 1s, then I suggest we:

* catch the `IOException` thrown by `waitNamedPipe`
* retry only `10` times or so, the last time we throw an exception.

> We don't know if this would block shutdown until it times out. If not it's ok.
I know, later we'll be able to check / test this.